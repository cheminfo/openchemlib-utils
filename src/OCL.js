let OCL;

export function getOCL() {
  if (!OCL) {
    throw new Error('OCL has to be initialized using initOCL(OCL)');
  }
  return OCL;
}
/**
 * 
 * @param {OCL} newOCL 
 * @param {object} [options={}]
 * @param {boolean} [options.keepExisting=false] Will keep current OCL if we initialize more than once
 */
export function initOCL(newOCL, options = {}) {
  const { keepExisting = false } = options;
  if (OCL && keepExisting) return;
  if (OCL && OCL !== newOCL) {
    throw new Error('OCL-utils was already initialized with a different OCL');
  }
  OCL = newOCL;
}
