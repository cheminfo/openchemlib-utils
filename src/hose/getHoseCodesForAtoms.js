import { isCsp3 } from '../util/isCsp3';
import { makeRacemic } from '../util/makeRacemic';
import { tagAtom } from '../util/tagAtom';

export const FULL_HOSE_CODE = 1;
export const HOSE_CODE_CUT_C_SP3_SP3 = 2;

/**
 * Returns the hose code for a specific atom number
 * @param {OCL.Molecule} originalMolecule
 * @param {array<number>} rootAtoms
 * @param {object} [options={}]
 * @param {boolean} [options.isTagged] Specify is the atoms are already tagged
 * @param {number} [options.minSphereSize=0] Smallest hose code sphere
 * @param {number} [options.maxSphereSize=4] Largest hose code sphere
 * @param {number} [options.kind=FULL_HOSE_CODE] Kind of hose code, default usual sphere
 */
export function getHoseCodesForAtoms(
  originalMolecule,
  rootAtoms = [],
  options = {},
) {
  const OCL = originalMolecule.getOCL();
  const {
    minSphereSize = 0,
    maxSphereSize = 4,
    kind = FULL_HOSE_CODE,
    isTagged = false,
  } = options;

  const molecule = originalMolecule.getCompactCopy();

  if (!isTagged) {
    const tags = [];
    for (let i = 0; i < rootAtoms.length; i++) {
      let rootAtom = rootAtoms[i];
      tags.push(tagAtom(molecule, rootAtom));
      molecule.addImplicitHydrogens();
      molecule.addMissingChirality();
      molecule.ensureHelperArrays(OCL.Molecule.cHelperNeighbours);
      // because ensuring helper reorder atoms we need to look again for it
    }
    rootAtoms.length = 0;
    for (let j = 0; j < molecule.getAllAtoms(); j++) {
      if (tags.includes(molecule.getAtomCustomLabel(j))) {
        rootAtoms.push(j);
      }
    }
  }

  let fragment = new OCL.Molecule(0, 0);
  let results = [];
  let min = 0;
  let max = 0;
  let atomMask = new Array(molecule.getAllAtoms());
  let atomList = new Array(molecule.getAllAtoms());

  for (let sphere = 0; sphere <= maxSphereSize; sphere++) {
    if (max === 0) {
      for (let rootAtom of rootAtoms) {
        atomList[max] = rootAtom;
        atomMask[rootAtom] = true;
        max++;
      }
    } else {
      let newMax = max;
      for (let i = min; i < max; i++) {
        let atom = atomList[i];
        for (let j = 0; j < molecule.getAllConnAtoms(atom); j++) {
          let connAtom = molecule.getConnAtom(atom, j);
          if (!atomMask[connAtom]) {
            switch (kind) {
              case FULL_HOSE_CODE:
                atomMask[connAtom] = true;
                atomList[newMax++] = connAtom;
                break;
              case HOSE_CODE_CUT_C_SP3_SP3:
                if (!(isCsp3(molecule, atom) && isCsp3(molecule, connAtom))) {
                  atomMask[connAtom] = true;
                  atomList[newMax++] = connAtom;
                }
                break;
              default:
                throw new Error('getHoseCoesForAtom unknown kind');
            }
          }
        }
      }
      min = max;
      max = newMax;
    }
    molecule.copyMoleculeByAtoms(fragment, atomMask, true, null);
    if (sphere >= minSphereSize) {
      makeRacemic(fragment);
      results.push(
        fragment.getCanonizedIDCode(
          OCL.Molecule.CANONIZER_ENCODE_ATOM_CUSTOM_LABELS,
        ),
      );
    }
  }
  return results;
}
