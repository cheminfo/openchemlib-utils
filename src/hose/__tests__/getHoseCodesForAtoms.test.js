import OCL from 'openchemlib';

import { getHoseCodesForAtoms } from '../getHoseCodesForAtoms';

const { Molecule } = OCL;

describe('getHoseCodesForAtoms', () => {
  it('C*C(Cl)CC single marked atom', () => {
    let molecule = Molecule.fromSmiles('CC(Cl)CC');
    let hoses = getHoseCodesForAtoms(molecule, [0]);
    hoses = hoses.map((hose) => escape(hose));
    expect(hoses).toStrictEqual([
      'fH@NJ%60%7FRapj%60',
      'eF@Hp%5CQP%5EcotdGBj@',
      'gC%60HADIMUIPNHjCrdKotbIpj%60',
      'gJPHADIMuTe@xbhOZPnv_iDSaU@',
      'gJPHADIMuTe@XbhOtbIpj%60',
    ]);
  });
  it('C*C*(Cl)CC double marked atoms', () => {
    let molecule = Molecule.fromSmiles('CC(Cl)CC');
    let hoses = getHoseCodesForAtoms(molecule, [0, 1]);
    hoses = hoses.map((hose) => escape(hose));
    expect(hoses).toStrictEqual([
      'eF@Hp%5CIPUA%7Eb%60xUTxUP',
      'gC%60HADIMUIPNEJATGeHW_hdcaUIpj%60',
      'gJPHADIMuTe@xThEP%5Eta%5Dl%7FQIGBjSaU@',
      'gJPHADIMuTe@XThEP_hdcaUIpj%60',
      'gJPHADIMuTe@XThEP_hdcaUIpj%60',
    ]);
  });
  it('Hc1ccccc1 double marked atoms in aromatic', () => {
    let molecule = Molecule.fromSmiles('Cc1ccccc1');
    molecule.setAtomicNo(0, 1);
    molecule.addImplicitHydrogens();
    const results = [];
    for (let i = 0; i < molecule.getAllAtoms(); i++) {
      results.push(
        getHoseCodesForAtoms(molecule, [i], { maxSphereSize: 0 })[0],
      );
    }
    expect(results).toStrictEqual([
      'fH@NJ`\x7FRapj`',
      'fH@NJ`\x7FRapj`',
      'fH@NJ`\x7FRapj`',
      'fH@NJ`\x7FRapj`',
      'fH@NJ`\x7FRapj`',
      'fH@NJ`\x7FRapj`',
      'fHdrA\x7FRaDj`',
      'fHdrA\x7FRaDj`',
      'fHdrA\x7FRaDj`',
      'fHdrA\x7FRaDj`',
      'fHdrA\x7FRaDj`',
      'fHdrA\x7FRaDj`',
    ]);
  });
});
