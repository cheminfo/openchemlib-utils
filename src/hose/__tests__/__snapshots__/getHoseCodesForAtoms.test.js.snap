// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getHoseCodesForAtoms cyclosporin 1`] = `
[
  "fH@NJ\`Rapj\`",
  "gCh@@duPGDeAzROtbQpj\`",
  "fJ@GEP_iPnUP",
  "gCh@AGj@x\`hOtbA\\j\`",
  "gCi@DDfZ@xdhOtbQpj\`",
  "gCa@@duPGDeAzROtbQpj\`",
  "eF\`BLGETGj{}IIpj\`",
  "eM\`BN\`xb\`}D_iHKeT",
  "fI@GEP_iP~UP",
  "eFHBJGDTGzRCyU@",
  "gC\`@H}PGDEAzBOtbApj\`",
  "eFHBLGDTGhc}IA|j\`",
  "eM@HzCbJCtY~d\`xUP",
  "eF@Hp\\QP^cotdGBj@",
  "eM@HvCbJCtQ~d\`xUP",
  "eM\`AIhNJhOUgzRSaU@",
  "fHdrARaDj\`",
  "eFBBYcARPQJh",
  "eFbHfXp_tdDRj@",
  "eFJHfXp_tdDRj@",
]
`;
